internal/resources/gitrepository/data_source_git_repository.go

```diff
@@ -5,10 +5,192 @@ SPDX-License-Identifier: MPL-2.0
+	if err != nil {
+		if clienterrors.IsNotFoundError(err) && !helper.IsDataRead(ctx) {
+			_ = schema.RemoveFromState(d, m)
+			return
+		}
+
+		return diag.FromErr(err)
+	}
+
+	// always run
+	d.SetId(UID)
+
+	if err := d.Set(common.MetaKey, common.FlattenMeta(meta)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	var (
+		flattenedSpec   []interface{}
+		flattenedStatus interface{}
+	)
+
+	switch scopedFullnameData.Scope {
+	case commonscope.ClusterScope:
+		flattenedSpec = spec.FlattenSpecForClusterScope(atomicSpec)
+		flattenedStatus = status.FlattenStatusForClusterScope(clusterScopeStatus)
+	case commonscope.ClusterGroupScope:
+		clusterGroupScopeSpec := &gitrepositoryclustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupNamespaceFluxcdGitrepositorySpec{
+			AtomicSpec: atomicSpec,
+		}
+		flattenedSpec = spec.FlattenSpecForClusterGroupScope(clusterGroupScopeSpec)
+		flattenedStatus = status.FlattenStatusForClusterGroupScope(clusterGroupScopeStatus)
+	}
+
+	if err := d.Set(spec.SpecKey, flattenedSpec); err != nil {
+		return diag.FromErr(err)
 	}
+
+	if err := d.Set(statusKey, flattenedStatus); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+// nolint: gocognit
+func retrieveGitRepositoryUIDMetaAndSpecFromServer(config authctx.TanzuContext, scopedFullnameData *scope.ScopedFullname, d *schema.ResourceData) (
+	string,
+	*objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta,
+	*gitrepositoryclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdGitrepositorySpec,
+	*gitrepositoryclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdGitrepositoryStatus,
+	*gitrepositoryclustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupNamespaceFluxcdGitrepositoryStatus,
+	error) {
+	var (
+		UID                     string
+		meta                    *objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta
+		spec                    *gitrepositoryclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdGitrepositorySpec
+		clusterScopeStatus      *gitrepositoryclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdGitrepositoryStatus
+		clusterGroupScopeStatus *gitrepositoryclustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupNamespaceFluxcdGitrepositoryStatus
+	)
+	// nolint: dupl
+	switch scopedFullnameData.Scope {
+	case commonscope.ClusterScope:
+		if scopedFullnameData.FullnameCluster != nil {
+			resp, err := config.TMCConnection.ClusterGitRepositoryResourceService.VmwareTanzuManageV1alpha1ClusterFluxcdGitrepositoryResourceServiceGet(scopedFullnameData.FullnameCluster)
+			if err != nil {
+				if clienterrors.IsNotFoundError(err) {
+					d.SetId("")
+					return "", nil, nil, nil, nil, err
+				}
+
+				return "", nil, nil, nil, nil, errors.Wrapf(err, "Unable to get Tanzu Mission Control cluster git repository entry, name : %s", scopedFullnameData.FullnameCluster.Name)
+			}
+
+			scopedFullnameData = &scope.ScopedFullname{
```

Comment:
code duplication between the cases can be generalised and reduced, can it be taken up ?